tests/unit/test_task_outputs.py

```diff
@@ -13,30 +13,297 @@
+            'succeeded or submit_failed or expired',
+            id='5',
+        ),
+        pytest.param(
+            {TASK_OUTPUT_SUCCEEDED, TASK_OUTPUT_SUBMITTED},
+            {TASK_OUTPUT_EXPIRED},
+            '(submitted and succeeded) or expired',
+            id='6',
+        ),
+        pytest.param(
+            [],
+            {TASK_OUTPUT_SUCCEEDED, TASK_OUTPUT_SUBMIT_FAILED},
+            'succeeded or failed or submit_failed',
+            id='7',
+        ),
+        pytest.param(
+            {'x'},
+            {
+                TASK_OUTPUT_SUCCEEDED,
+                TASK_OUTPUT_SUBMIT_FAILED,
+                TASK_OUTPUT_EXPIRED,
+            },
+            '(x and succeeded) or failed or submit_failed or expired',
+            id='8',
+        ),
+    ],
+)
+def test_get_completion_expression_implicit(required, optional, expression):
+    """It should generate a completion expression if none is provided."""
+    assert get_completion_expression(tdef(required, optional)) == expression
+
+
+def test_get_completion_expression_explicit():
+    """If a completion expression is used, it should be used unmodified."""
+    assert get_completion_expression(tdef(
+        {'x', 'y'},
+        {TASK_OUTPUT_SUCCEEDED, TASK_OUTPUT_FAILED, TASK_OUTPUT_EXPIRED},
+        '((failed and x) or (succeeded and y)) or expired'
+    )) == '((failed and x) or (succeeded and y)) or expired'
+
+
+def test_format_completion_status():
+    outputs = TaskOutputs(
+        tdef(
+            {TASK_OUTPUT_SUCCEEDED, 'x', 'y'},
+            {TASK_OUTPUT_EXPIRED},
+        )
+    )
+    assert outputs.format_completion_status(
+        indent=2, gutter=2
+    ) == '  ' + sstrip(
+        '''
+          ⦙  (
+        ⨯ ⦙    succeeded
+        ⨯ ⦙    and x
+        ⨯ ⦙    and y
+          ⦙  )
+        ⨯ ⦙  or expired
+        '''
+    )
+    outputs.set_message_complete('succeeded')
+    outputs.set_message_complete('x')
+    assert outputs.format_completion_status(
+        indent=2, gutter=2
+    ) == '  ' + sstrip(
+        '''
+          ⦙  (
+        ✓ ⦙    succeeded
+        ✓ ⦙    and x
+        ⨯ ⦙    and y
+          ⦙  )
+        ⨯ ⦙  or expired
+        '''
+    )
+
+
+def test_iter_required_outputs():
+    """It should yield required outputs only."""
+    # this task has three required outputs and one optional output
+    outputs = TaskOutputs(
+        tdef(
+            {TASK_OUTPUT_SUCCEEDED, 'x', 'y'},
+            {'z'}
+        )
+    )
+    assert set(outputs.iter_required_messages()) == {
+        TASK_OUTPUT_SUCCEEDED,
+        'x',
+        'y',
+    }
+
+    # this task does not have any required outputs (besides the implicitly
+    # required submitted/started outputs)
+    outputs = TaskOutputs(
+        tdef(
+            {TASK_OUTPUT_SUCCEEDED, 'x', 'y'},
+            {TASK_OUTPUT_FAILED},  # task may fail
+        )
+    )
+    assert set(outputs.iter_required_messages()) == set()
```

Comment:
It's confusing that this `tdef(required={'succeeded', 'x', 'y'}, ...` results in `succeeded`, `x` and `y` being required if `optional={'z'}`, but not being required if `optional={'failed'}`

Reply:
This is as per the spec: https://cylc.github.io/cylc-admin/proposal-optional-output-extension.html

If a task succeeds then it is expected to generate its required outputs, if it fails (and failure is permitted), then it is not.

Reply:
What does the graph for this look like? It seems to me that this doesn't make sense and it should be
```suggestion
    # this task does not have any required outputs (besides the implicitly
    # required submitted/started outputs)
    outputs = TaskOutputs(
        tdef(
            {},
            {TASK_OUTPUT_SUCCEEDED, 'x', 'y', TASK_OUTPUT_FAILED},  # task may fail
        )
    )
    assert set(outputs.iter_required_messages()) == set()
```

Reply:
It doesn't, this is a unit test probing the `TaskOutputs` stuff directly to ensure consistency at that level. This exact example wouldn't actually get through validation (failure cannot be optional if success is required). If you replace `TASK_OUTPUT_FAILED` with `TASK_OUTPUT_EXPIRED` then validation won't complain and this could be defined in a graph along the lines of:

```
a
a:x
a:y
a:expired?
```

Purpose of the test is to shows that failure can bypass the required outputs if that's what the user wants.

Reply:
Added a comment.